packages/block-editor/src/components/block-list/block.native.js

```diff
@@ -187,18 +258,33 @@ export default compose( [
 		const icon = blockType.icon;
 		const getAccessibilityLabelExtra = blockType.__experimentalGetAccessibilityLabel;
 
-		const selectedBlock = getSelectedBlock();
-		const parentId = getBlockRootClientId( clientId );
-		const parentBlock = getBlock( parentId );
-
-		const isMediaText = selectedBlock && selectedBlock.name === 'core/media-text';
-		const isMediaTextParent = parentBlock && parentBlock.name === 'core/media-text';
+		const parents = getBlockParents( clientId, true );
+		const parentId = parents[ 0 ] || '';
 
 		const rootBlockId = getBlockHierarchyRootClientId( clientId );
 		const rootBlock = getBlock( rootBlockId );
 		const hasRootInnerBlocks = rootBlock.innerBlocks.length !== 0;
 
-		const showFloatingToolbar = isSelected && hasRootInnerBlocks && ! isMediaText && ! isMediaTextParent;
+		const showFloatingToolbar = isSelected && hasRootInnerBlocks;
+
+		const selectedBlockClientId = getSelectedBlockClientId();
+
+		const commonAncestor = getLowestCommonAncestorWithSelectedBlock( clientId );
+		const commonAncestorIndex = parents.indexOf( commonAncestor ) - 1;
+		const firstToSelectId = commonAncestor ? parents[ commonAncestorIndex ] : parents[ parents.length - 1 ];
+
+		const hasChildren = !! getBlockCount( clientId );
+		const hasParent = !! parents[ 0 ];
+		const isParentSelected = selectedBlockClientId && selectedBlockClientId === parents[ 0 ];
+		const isAncestorSelected = selectedBlockClientId && parents.includes( selectedBlockClientId );
+		const selectionIsNested = !! getBlockRootClientId( selectedBlockClientId );
+
+		const isDescendantSelected = selectedBlockClientId && getBlockParents( selectedBlockClientId ).includes( clientId );
+		const isDescendantOfParentSelected = selectedBlockClientId && getBlockParents( selectedBlockClientId ).includes( parentId );
+		const isTouchable = isSelected || isDescendantSelected || isDescendantOfParentSelected || isParentSelected || parentId === '';
+		const isDimmed = ! isSelected && selectionIsNested && ! isAncestorSelected && ! isDescendantSelected && ( isDescendantOfParentSelected || rootBlockId === clientId );
```

Comment:
Those conditions looks  quite hard to understand! Could we try to make them more readable? I haven't spent that much time trying to understand it yet so I might have omitted some edge cases, but from the few cases I can think of, the following conditions would be more readable imo:
```
const isBlockHighlighted = selectionIsNested && ( isSelected || isParentSelected );
const isDimmed =  ! isBlockHighlighted;
```
I'm not sure how `isDescendantOfParentSelected` could play a role here, is there a case where a component is highlighted when a "sibling" is selected?

Reply:
As I understand, `isDimmed` should only be true for blocks that are outside the selection and its descendants, right? There was a suggestion to do a different think, but as far as I understand we didn't decide to go that way. It's not clear to me if we want to dim other blocks when a top level one is selected (even if it has children), or just when we start going 1+ level deep into nested blocks.

There are some redundant checks:
- If `isDescendantSelected` ⇒ `isDescendantOfParentSelected`. So `isDescendantSelected || isDescendantOfParentSelected == isDescendantSelected`

Reply:
>I'm not sure how isDescendantOfParentSelected could play a role here, is there a case where a component is highlighted when a "sibling" is selected?

In my version it's the case when you select first child of root block. Without this check the rest of the components stays highlighted.

Next thing is that I need to dim only the bottom most descendant in each branch which should be dimmed to avoid multiplication of opacity applied to block in hierarchy. In other words if I apply opacity to block parent which fits criteria to be dimmed and them apply opacity for the block itself the block will have opacity 0.4 instead of desire 0.2. This opacity increased depending of nesting level. I hope that make sense

There is also edge case that do not dim when RootList block is selected as well.

I gave a shot for what you proposed and it does not works as expected but I will try to refactor original checks to increase clarity and write some comments in the code

Reply:
@koke I have noticed your comment now so I want to refer to what you wrote as well

>There are some redundant checks:
If isDescendantSelected ⇒ isDescendantOfParentSelected. So isDescendantSelected || isDescendantOfParentSelected == isDescendantSelected

That's true according to same branch.

The case for check 'isDescendantOfParentSelected' according to below image is when we have **`1-4`** selected  (red) and `isTouchable` should be set to true for **`1-2`**, **`1-5`**, **`1-4-5`** ,**`1-4-3`** (green). The blue one should be dimmed.

For **`1-5`** this statement `isDescendantSelected || isDescendantOfParentSelected == isDescendantSelected` is not true because **`1-4`** _`is not`_ descendant of `1-5` but _`is`_ descendant of **`1-5`** parent

I also agree that we can simplify this:
`const isTouchable = isSelected || isDescendantSelected || isDescendantOfParentSelected || isParentSelected || parentId === '';
`
to this:
`const isTouchable =  isSelected || isDescendantOfParentSelected || isParentSelected || parentId === '';`

<image src="https://user-images.githubusercontent.com/21242757/69862172-24ce8800-129a-11ea-8aff-d5c2dc242b8c.png" width="300" />

Reply:
That's right, I knew one of them was redundant but I got my logic backwards 😅

Reply:
I'm still not sure about `isDimmed` and try to check if we can simplify